### PR TITLE
Reorganize project and add initial Diagnostic API

### PR DESCRIPTION
**Pull request by harlanhaskins**
This adds an initial Diagnostics API that we can use while building the parser, but I would like to better architect the notion of diagnostic messages. Right now, they're just `String`s, which feels Bad™. I'd love to design an API where a specific diagnostic is an object that takes parameters and knows how to format a message automatically.

@CodaFi What are your thoughts on the API as-is, and what do you suggest for a better architecture?

### CODE REVIEW

**Review by CodaFi — APPROVED** [review]
We had a lot of offline discussion and arrived at this API

**Comment by CodaFi** [comment]
Also, experimenting with typelift/SwiftCheck#253 to see if we can get this transitive dependency crap resolved.

**Comment by harlanhaskins** [comment]
Once this passes, I'll merge

File: Package.swift [hunk 1]
```diff
@@ -4,28 +4,38 @@
...
+      dependencies: ["Rainbow"]),
+    .target(
+      name: "Crust",
+      dependencies: ["Lithosphere"]),
+    .target(
+      name: "Drill",
+      dependencies: ["Lithosphere", "Crust"]),
+    .target(
+      name: "silt",
+      dependencies: ["Drill"]),
```
Comment (CodaFi) [inline Package.swift]: `silt` depends on `CommandLineKit`

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**